### PR TITLE
Fix HepMC3 reading of vertices and status codes

### PR DESCRIPTION
**Pull request by sethrj**
Follow-on to #854: previously Celeritas only looked at the "event position" rather than the vertex position in a HepMC3 event. It also provided as primary particles *all* tracks with positive status codes, although in practice only those with status code of 1 are meant to be transported in detector simulation. This fixes reading for both EventReader and the HepMC3 vertex generator.

I think we might be abusing the HepMC3 format: I'm guessing the initial tracks (the ones with the "event"/default vertex) are supposed to be only the incident collider beam particles (not to be simulated), and all the subsequent vertices are allowed to have status code 1. In [the HepMCG4Interface.cc example]( https://github.com/Geant4/geant4/blob/dd1f179cda58f54140945ad67846ff417903a862/examples/extended/eventgenerator/HepMC/HepMCEx01/src/HepMCG4Interface.cc) none of the "event-level" particles are iterated over, so for example the HepMC3 file generated by @stognini 's [hepmc3-generator](https://github.com/celeritas-project/utils/tree/master/hepmc3-generator) won't show up as primaries in that Geant4 example because they don't belong to a vertex.

### CODE REVIEW

**Comment by sethrj** [comment]
@amandalund I think this is finally ready for review...

File: app/data/ttbarbits-12evt-118prim.hepmc3 [hunk 1]
```diff
@@ -2,93 +2,93 @@ HepMC::Version 3.02.05
 HepMC::Asciiv3-START_EVENT_LISTING
 E 1 0 10
 U GEV MM
-P 1 0 -24 -4.7100161999999997e+01 -1.2502119999999999e+02 -2.2036460000000000e+02 2.6993799103269066e+02 8.0356931000000003e+01 2
-P 2 0 24 1.5931562000000000e+02 -2.8732734000000001e+01 -2.1139886999999999e+02 2.7794694522156675e+02 7.9736975999999999e+01 2
-P 3 0 -24 -4.5609923999999999e+01 -1.1829483999999999e+02 -2.0798566000000000e+02 2.5649406516191704e+02 8.0356931000000003e+01 2
-P 4 0 24 1.5891660999999999e+02 -2.8659172999999999e+01 -2.1094170000000000e+02 2.7736298107615335e+02 7.9736975999999999e+01 2
+P 1 0 -24 -4.7100161999999997e+01 -1.2502119999999999e+02 -2.2036460000000000e+02 2.6993799103269066e+02 8.0356931000000003e+01 1
+P 2 0 24 1.5931562000000000e+02 -2.8732734000000001e+01 -2.1139886999999999e+02 2.7794694522156675e+02 7.9736975999999999e+01 1
```
Comment (amandalund) [inline app/data/ttbarbits-12evt-118prim.hepmc3:6]: I guess none of these had an outgoing vertex? Did we generate this file?

**Review by amandalund — APPROVED** [review]
Thanks for the fixes @sethrj!

Probably would have helped us earlier, but for future reference from the [HepMC3 paper](https://doi.org/10.1016/j.cpc.2020.107310):

> The particles that act as graph edges between vertices typically have a "production vertex” where they came from, and an “end vertex”, where they undergo their next modification or interaction: the only exceptions to this rule are the stable final-state particles which have no end vertex, with the (usually two) incoming beam particles which are assigned to a unique “root vertex” without incoming particles.
>
> The HepMC3 event record can hold events with arbitrarily complex relations between the particles and vertices. However, to avoid algorithmic problems, it is expected the event structure will adhere to the following rules:
> - All particles and vertices in the event should be connected with each other, e.g. the event should not contain dangling particles or vertices.
> - Cyclic relations where a particle can be its own ancestor should be avoided.
> - All vertices should have at least one outgoing particle.
> - All vertices but root vertex should have at least one incoming particle.
> - Vertices should have a meaningful or zero status code:
>   - Particles with no end vertex should be assigned status 1;
>   - The incoming particles should be assigned status 4.

**Comment by sethrj** [comment]
Ugh, of *course* their documentation would be a publication and not on their code documentation page.

File: app/data/ttbarbits-12evt-118prim.hepmc3 (hunk 1 above)
Comment (sethrj) [reply app/data/ttbarbits-12evt-118prim.hepmc3:6]: I made this a long time ago by taking some Pythia output file (which I think @stognini generated?) and trimmed out everything but the EM tracks... it's basically garbage. It lets the tests "run" ... if we want them to be meaningful we have to change a lot more than just the input file.

**Pull request merged**